File: ruby/medium/inorder_successor_in_bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {TreeNode} p
# @return {TreeNode}
def inorder_successor(root, p)
    res = {}
    find_inorder_successor(root, p, res)
    return res ? res[:successor] : nil
end

def find_inorder_successor(root, p, res)
    return unless root
    if(root == p)
        root = root.right
        while(root && root.left)
            root = root.left
        end
        res[:successor] = root
    elsif(p.val < root.val)
        find_inorder_successor(root.left, p, res)
        res[:successor] = root unless res[:successor]
    else
        find_inorder_successor(root.right, p ,res)
    end
end

File: javascript/medium/inorderSuccessorInBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @return {TreeNode}
 */
var inorderSuccessor = function(root, p) {
    const res = {};
    findInOrderSuccessor(root, p, res);
    return res.successor;
};

var findInOrderSuccessor = function(root, p, res) {
    if(root === null) return;
    if(root === p) {
        root = root.right;
        while(root !== null && root.left !== null) {
            root = root.left;
        }
        res.successor = root;
    } else if(p.val < root.val) {
        findInOrderSuccessor(root.left, p, res);
        if(res.successor === null) {
            res.successor = root;
        }
    } else {
        findInOrderSuccessor(root.right, p, res);
    }
};
